### ISSUE TITLE
[BUG][WebGL] Skill level in WebGL client is always default instead of whats in server/database

### ISSUE THREAD

**Issue by shubhank008** [suriyun-production/mmorpg-kit-docs#1911]
@insthync sorry for tagging but critical bug in webgl the level of the skill is being set/read as default level, instead of current level in database/server.

To explain:
I have a starting Skill for my character, its set to level 0 (correct in webgl and server)
Now I levelled up my skill to level 5 by playing (correct in webgl and server)
But if I open the game again in webgl, the skill level is back to level 0 in webgl (which I assume is the default level of the skill from my scriptable).

In both mysql database and server, skill level shows as level 5 properly.
If I run the game in Unity editor, skill shows properly as level 5.
If I run the game in webgl client, skill shows as level 0.

MySQL Database
![MobaXterm_Lsn65tDGVz](https://user-images.githubusercontent.com/3790163/218801374-143c28c8-b863-4170-a3bc-9e10cc620717.jpg)

To test and try to debug, I added these debug logs in MakeCache method of CharacterSkill class
![devenv_jLa07iulBX](https://user-images.githubusercontent.com/3790163/218802077-609272fe-b3eb-4f71-9ab3-d479ceb29643.jpg)

On Server (MMO), it reports the level properly as 5 (`-495378242 is the skill (dataId)`)
![MobaXterm_JxYyzVUeDr](https://user-images.githubusercontent.com/3790163/218801530-ba71d9ed-fba1-4a4c-8031-ae5d73db7b2b.jpg)

In WebGL it reports level 0
![chrome_9nmJTFszgp](https://user-images.githubusercontent.com/3790163/218801679-bab2db9f-a937-4c4f-bd04-0bb3341c9d1c.jpg)

No errors or anything in web browser console log or anything.

**Comment by shubhank008** [comment]
@insthync 

I did some extensive debugging and found a few things

```
using LiteNetLib.Utils;
using LiteNetLibManager;
using UnityEngine;

namespace MultiplayerARPG
{
    [System.Serializable]
    public class CharacterSkill : INetSerializable
    {
        public static readonly CharacterSkill Empty = new CharacterSkill();
        public int dataId;
        public short level;

        [System.NonSerialized]
        private int dirtyDataId;
        [System.NonSerialized]
        private short dirtyLevel;

        [System.NonSerialized]
        private BaseSkill cacheSkill;

        private void MakeCache()
        {
            if (dirtyDataId != dataId || dirtyLevel != level)
            {                
                dirtyDataId = dataId;
                dirtyLevel = level;
                cacheSkill = null;
                GameInstance.Skills.TryGetValue(dataId, out cacheSkill);
                if (dataId == -495378242) { 
                    Debug.Log("CharacterSkill: " + dirtyDataId + " " + dataId + " " + dirtyLevel + " " + level);
                    if (cacheSkill != null)
                    {
                        Debug.Log("CharacterSkill: " + cacheSkill.name + " " + cacheSkill.Title);
                    }
                    else
                    {
                        Debug.Log("CharacterSkill: cacheSkill is null");
                    }
                }                
            }
        }

        public BaseSkill GetSkill()
        {
            MakeCache();
            return cacheSkill;
        }

        public CharacterSkill Clone()
        {
            if (dataId == -495378242)
            {
                Debug.Log("CharacterSkillCLONE: " + dataId + " " + level);
            }
            return new CharacterSkill()
            {
                dataId = dataId,
                level = level,
            };
        }

        public static CharacterSkill Create(BaseSkill skill, short level = 1)
        {
            if (skill.DataId == -495378242)
            {
                Debug.Log("CharacterSkillCreate: " + skill.DataId + " " + level);
            }
            return Create(skill.DataId, level);
        }

        public static CharacterSkill Create(int dataId, short level = 1)
        {
            if (dataId == -495378242)
            {
                Debug.Log("CharacterSkillCreate2: " + dataId + " " + level);
            }
            return new CharacterSkill()
            {
                dataId = dataId,
                level = level,
            };
        }

        public void Serialize(NetDataWriter writer)
        {
            if (dataId == -495378242)
            {
                Debug.Log("CharacterSkillSerialize: " + dataId + " " + level);
            }
            writer.PutPackedInt(dataId);
            writer.PutPackedShort(level);
        }

        public void Deserialize(NetDataReader reader)
        {
            dataId = reader.GetPackedInt();
            level = reader.GetPackedShort();

            if (dataId == -495378242)
            {
                Debug.Log("CharacterSkillDeserialize: " + dataId + " " + level);
            }
        }
    }

    [System.Serializable]
    public class SyncListCharacterSkill : LiteNetLibSyncList<CharacterSkill>
    {
    }
}
```

The deserialize method of characterskill is not called in webgl client, while its the first thing thats called in editor/standalone
Similarly `GetCharacterSkills()` from `CharacterDataExtensions` is also not called in webgl, while its the 2nd call in standalone

How my editor/standalone looks (debug prints skill.dataId and level)
![image](https://user-images.githubusercontent.com/3790163/218887018-84374cde-cca3-4438-ab89-72cff1f6f813.png)

This was how my webgl looked, similarly skill.dataId and level (which is 0, what I have set as starting skill level for the chara scriptable)
![image](https://user-images.githubusercontent.com/3790163/218887093-8b5ef249-443b-45f4-88b5-f0bf5a7bb177.png)


As you can see, in webgl first thing called is the `public static CharacterSkill Create()` method instead of deserialize or other calls like in standalone

**Comment by insthync** [comment]
On other platforms don't having this issues?

**Comment by insthync** [comment]
Asking because it is having validate function which will vanish skill list which is not in PlayerCharacter data, I think I will change it workflow.

**Comment by shubhank008** [comment]
> On other platforms don't having this issues?

No, works on windows and even in editor, webgl only problem
I checked and don't think its validate function problem more like the SyncList is never received/synced on webgl and thus the character skill data never deserializes

**Comment by shubhank008** [comment]
@insthync so upon further investigation, found out that in the 001Home scene in character selection screen, the SyncList works fine, characterSkill deserialize is called and it reports proper skill level (being 5)

But it is when you start game and spawn in map, that the SyncList doesn't seem to work, characterskill.deserialize is not called and the skill level stays default from scriptable (being 0)

**Comment by shubhank008** [comment]
@insthync any update ? this is quite critical as game breaking on webgl and preventing us to ship out our Beta

**Comment by insthync** [comment]
Try to fix it, really doubt why only character skill

**Comment by shubhank008** [comment]
> Currently trying to fix it, really doubt why only character skill

Don't know either, I tested items and skill as a item and they work fine, only skill level is bugged
And again, the level syncs fine till Home/001Home screen but on the map/after spawn it doesnt so its either something related to mapmanager or synclist not working, I spent last 2 days trying to get as much data I could to help you but litenetlib networking is out of my expertise so couldn't get anything further

Basically I was trying to check if the synclist is getting received on clientside

**Comment by insthync** [comment]
I can fix it by changing a variable name from `skills` to another

![Screenshot_373](https://user-images.githubusercontent.com/2697550/219877223-546208e7-f4c1-4970-8e19-b88b856be8e6.png)

**Comment by insthync** [comment]
Fixed
https://github.com/insthync/LiteNetLibManager/commit/39575885093fe7474b21a8a5885ece42b44ba74f

**Comment by shubhank008** [comment]
> Fixed [insthync/LiteNetLibManager@3957588](https://github.com/insthync/LiteNetLibManager/commit/39575885093fe7474b21a8a5885ece42b44ba74f)

Thanks, testing and reporting back.

If you don't mind me asking, out of curiosity how were you able to figure it out or nail it down ? So in future I can save as much time for you when reporting and debugging the problem.

**Comment by insthync** [comment]
In this case, I can guess what happened just because I know how data is synced, if other data can be synced then why just skills can't be synced properly so I guess that it might have been something wrong at the client because the skill sync list will be synced from server to client, clients on other platforms can receive data but only WebGL that having issues, then I try to write log into prepare and process sync field functions to find out that the index of sync list is correct or not and can find that it is not.

**Comment by shubhank008** [comment]
@insthync I can confirm the above patch fixes this bug, tested and confirmed working now in webgl
Thank you for the quick fix. 

> Fixed [insthync/LiteNetLibManager@3957588](https://github.com/insthync/LiteNetLibManager/commit/39575885093fe7474b21a8a5885ece42b44ba74f)
